p2p/pex/pex_reactor.go

```diff
@@ -453,47 +453,91 @@ func (r *Reactor) ensurePeers() {
 	// not perfect, but somewhate ensures that we prioritize connecting to more-vetted
 	// NOTE: range here is [10, 90]. Too high ?
 	newBias := cmtmath.MinInt(out, 8)*10 + 10
-
 	toDial := make(map[p2p.ID]*p2p.NetAddress)
-	// Try maxAttempts times to pick numToDial addresses to dial
-	maxAttempts := numToDial * 3
 
-	for i := 0; i < maxAttempts && len(toDial) < numToDial; i++ {
-		if !r.IsRunning() || !r.book.IsRunning() {
-			return
-		}
+	// Calculate reserve size as 50% of numToDial with a minimum of 10
+	reserveSize := cmtmath.MaxInt(numToDial/2, 10)
 
-		try := r.book.PickAddress(newBias)
-		if try == nil {
-			continue
-		}
-		if _, selected := toDial[try.ID]; selected {
+	// Try maxAttempts times to pick numToDial + reserveSize addresses to dial
+	maxToDialAttempts := numToDial * 3
+
+	for i := 0; i < maxToDialAttempts && len(toDial) < numToDial+reserveSize; i++ {
```

Comment:
I am trying to understand the logic of this for.

In the worst case, we run it `numToDial * 3` times.

Or we wait until `len(toDial)` to reach `numToDial+reserveSize`.

We add to `toDial` good peer candidates, according to the filtering method.

But how many peers does `r.book.GetSelectionWithBias(newBias)` return? I that up to 23% of the number of stored addresses, up to 250 addresses.